package/MDAnalysis/transformations/nojump.py

```diff
@@ -0,0 +1,127 @@
+#
+# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
+# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+# doi: 10.25080/majora-629e541a-00e
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+"""\
+No Jump Trajectory Unwrapping --- :mod:`MDAnalysis.transformations.nojump`
+=======================================================================================
+
+Unwraps the trajectory such that atoms never move more than half a periodic box length.
+The algorithm used is based on :cite:p:`Kulke2022`.
+
+.. autoclass:: NoJump
+
+"""
+import numpy as np
+import warnings
+
+from .base import TransformationBase
+from ..due import due, Doi
+
+
+class NoJump(TransformationBase):
+    """
+    Returns transformed coordinates for the given timestep so that an atom
+    does not move more than half the periodic box size, and will not jump
+    across the periodic boundary. The algorithm used is based on :cite:p:`Kulke2022`, 
+    equation B6 for non-orthogonal systems.
+
+    Example
+    -------
+
+    To calculate diffusion, one of the first steps is to compute a trajectory
+    where the atoms do not cross the periodic boundary.
+
+    .. code-block:: python
+
+        transformation = NoJump()
+        u.trajectory.add_transformations(transformation)
+        for ts in u.trajectory:
+            print(ts.positions)
+
+    In this case, ``ts.positions`` will return the NoJump unwrapped trajectory.
+    To reverse the process, wrap the coordinates again.
+
+    Returns
+    -------
+    MDAnalysis.coordinates.timestep.Timestep
+
+    References
+    ----------
+    .. bibliography::
+        :filter: False
+        :style: MDA
+
+        Kulke2022
+
+    """
+
+    @due.dcite(Doi("10.1021/acs.jctc.2c00327"),
+               description="Works through the orthogonal case, "
+               "and proposes the non-orthogonal approach.", path=__name__)
+
+    def __init__(
+        self,
+        check_continuity=True,
+        max_threads=None,
+        # NoJump transforms are inherently unparallelizable, since
+        # it depends on the previous frame's unwrapped coordinates
+        parallelizable=False,
+    ):
+        super().__init__(max_threads=max_threads, parallelizable=parallelizable)
+        self.prev = None
+        self.old_frame = 0
+        self.check_c = check_continuity
+
+    def _transform(self, ts):
+        if self.prev is None:
+            self.prev = ts.positions @ np.linalg.inv(ts.triclinic_dimensions)
+            self.old_frame = ts.frame
+            return ts
```

Comment:
How does this work when I go through the trajectory a second time:
```python
u.add_transformation(NoJump) 
[ts for ts in u.trajectory]          
[ts for ts in u.trajectory[2:]]     # once the trajectory resets, will prev point to the correct starting value?
```
I think there should be a test that it's still correct the second time around. Keeping state is tricky business and our readers' behavior is not always obvious (for instance, they should rewind when you start iteration).

Reply:
It makes ZERO sense to unwrap an already unwrapped trajectory, and so ideally the transformation should only ever act once... I think. I could adopt something similar to what positionaveraging does, and try and remember what direction the steps are going in. However, in principle, unwrapping an already unwrapped trajectory is basically a no-op, since the round term will always be zero. I suppose I'd just need to enact a reset.

Reply:
Note that the trajectory will not stay unwrapped. It’s not in memory, the coordinates are read from disk again. (Sorry if I misunderstood your comment… typing quickly before they really force me to switch to airplane mode)

Reply:
Only if the user is using the MemoryReader then changes are permanent but that’s handled automatically with transformations.

Reply:
Hmm, ok. Now it'll throw a warning that you are going in the opposite direction (or with an unequal stride). When it encounters this, it will reset the unwrapping procedure.